File: src/domain/Layout.ts
import type {
	Board,
	BoardNode,
	CardNode,
	FrameNode,
	CardFrontmatter,
	Viewport,
} from "../types";
import {
	CARD_W,
	CARD_H,
	CARD_GAP,
	FRAME_LABEL_H,
	AI_INBOX_LABEL,
} from "../types";

export function findFrameByLabel(board: Board, label: string): FrameNode | undefined {
	return board.nodes.find(
		(n): n is FrameNode => n.type === "frame" && n.label === label
	);
}

export function findSystemFrame(board: Board, label: string): FrameNode | undefined {
	return board.nodes.find(
		(n): n is FrameNode => n.type === "frame" && n.system === true && n.label === label
	);
}

export function getCardsOverlappingFrame(board: Board, frame: FrameNode): CardNode[] {
	return board.nodes.filter((n): n is CardNode => {
		if (n.type !== "card") return false;
		return (
			n.x >= frame.x &&
			n.y >= frame.y &&
			n.x + n.w <= frame.x + frame.w &&
			n.y + n.h <= frame.y + frame.h
		);
	});
}

export function placeInFrame(board: Board, frame: FrameNode, cardId: string): CardNode {
	const existing = getCardsOverlappingFrame(board, frame);
	const cols = Math.max(1, Math.floor((frame.w - CARD_GAP) / (CARD_W + CARD_GAP)));
	const idx = existing.length;
	const col = idx % cols;
	const row = Math.floor(idx / cols);

	const neededH = FRAME_LABEL_H + CARD_GAP + (row + 1) * (CARD_H + CARD_GAP);
	if (neededH > frame.h) {
		frame.h = neededH;
	}

	return {
		id: `n_${cardId}`,
		type: "card",
		card_id: cardId,
		x: frame.x + CARD_GAP + col * (CARD_W + CARD_GAP),
		y: frame.y + FRAME_LABEL_H + CARD_GAP + row * (CARD_H + CARD_GAP),
		w: CARD_W,
		h: CARD_H,
		z: 1,
	};
}

export function createFrameNearViewport(viewport: Viewport, label: string): FrameNode {
	const x = viewport.x + Math.round(1200 / viewport.zoom);
	const y = viewport.y;
	return {
		id: `n_frame_${Date.now()}`,
		type: "frame",
		label,
		x,
		y,
		w: 1000,
		h: 600,
		color: "#88aaff",
	};
}

export function placeNewCard(
	board: Board,
	card: CardFrontmatter,
	group?: string
): { node: CardNode; newFrame?: FrameNode } {
	let targetFrame: FrameNode | undefined;
	let newFrame: FrameNode | undefined;

	if (group) {
		targetFrame = findFrameByLabel(board, group);
		if (!targetFrame) {
			newFrame = createFrameNearViewport(board.viewport, group);
			board.nodes.push(newFrame);
			targetFrame = newFrame;
		}
	} else {
		targetFrame = findSystemFrame(board, AI_INBOX_LABEL);
	}

	if (!targetFrame) {
		throw new Error(`AI Inbox frame not found in board`);
	}

	const node = placeInFrame(board, targetFrame, card.id);
	return { node, newFrame };
}

export function placeCenterViewport(viewport: Viewport, cardId: string): CardNode {
	return {
		id: `n_${cardId}`,
		type: "card",
		card_id: cardId,
		x: viewport.x + Math.round((800 - CARD_W) / 2 / viewport.zoom),
		y: viewport.y + Math.round((600 - CARD_H) / 2 / viewport.zoom),
		w: CARD_W,
		h: CARD_H,
		z: 1,
	};
}

export function ensureSystemFrames(board: Board): Board {
	const hasInbox = board.nodes.some(
		(n): n is FrameNode =>
			n.type === "frame" && n.system === true && n.label === AI_INBOX_LABEL
	);
	if (!hasInbox) {
		board.nodes.push({
			id: "n_inbox",
			type: "frame",
			label: AI_INBOX_LABEL,
			x: -1200,
			y: -400,
			w: 1000,
			h: 600,
			color: "#6688cc",
			system: true,
		});
	}
	return board;
}

File: link-index-mcp/src/layout.rs
use crate::board::{Board, BoardNode, Viewport};

const CARD_W: f64 = 280.0;
const CARD_H: f64 = 200.0;
const CARD_GAP: f64 = 16.0;
const FRAME_LABEL_H: f64 = 32.0;

/// Place a new card in the target frame (creating one if `group` is new).
/// Mutates `board.nodes` in place (may grow the frame height).
/// Returns the new CardNode to push onto `board.nodes`.
pub fn place_new_card(board: &mut Board, card_id: &str, group: Option<&str>) -> BoardNode {
    ensure_frame(board, group);
    place_in_frame(board, card_id, group)
}

/// Ensure a frame with the given label exists.
/// `group = None` targets the AI Inbox system frame (always present after `ensure_system_frames`).
fn ensure_frame(board: &mut Board, group: Option<&str>) {
    if let Some(g) = group {
        let exists = board.nodes.iter().any(|n| {
            matches!(n, BoardNode::Frame { label, .. } if label == g)
        });
        if !exists {
            let viewport = board.viewport.clone();
            board.nodes.push(new_frame(&viewport, g));
        }
    }
}

fn new_frame(vp: &Viewport, label: &str) -> BoardNode {
    BoardNode::Frame {
        id: format!("n_frame_{}", chrono::Utc::now().timestamp_millis()),
        label: label.to_string(),
        x: vp.x + (1200.0 / vp.zoom).round(),
        y: vp.y,
        w: 1000.0,
        h: 600.0,
        color: "#88aaff".to_string(),
        system: false,
    }
}

fn place_in_frame(board: &mut Board, card_id: &str, group: Option<&str>) -> BoardNode {
    // Find target frame geometry (immutable borrow ends after this block)
    let (frame_idx, fx, fy, fw, fh) = board
        .nodes
        .iter()
        .enumerate()
        .find_map(|(i, n)| {
            if let BoardNode::Frame { x, y, w, h, label, system, .. } = n {
                let matched = match group {
                    Some(g) => label == g,
                    None    => label == "AI Inbox" && *system,
                };
                if matched {
                    return Some((i, *x, *y, *w, *h));
                }
            }
            None
        })
        .expect("target frame must exist after ensure_frame");

    // Count cards overlapping the target frame
    let count = board.nodes.iter().filter(|n| {
        if let BoardNode::Card { x, y, w, h, .. } = n {
            *x >= fx && *y >= fy && (*x + *w) <= (fx + fw) && (*y + *h) <= (fy + fh)
        } else {
            false
        }
    }).count();

    // Compute grid position
    let cols = ((fw - CARD_GAP) / (CARD_W + CARD_GAP)).max(1.0).floor() as usize;
    let col  = count % cols;
    let row  = count / cols;

    // Expand frame height if needed
    let needed_h = FRAME_LABEL_H + CARD_GAP + (row + 1) as f64 * (CARD_H + CARD_GAP);
    if needed_h > fh {
        if let BoardNode::Frame { h, .. } = &mut board.nodes[frame_idx] {
            *h = needed_h;
        }
    }

    BoardNode::Card {
        id:      format!("n_{card_id}"),
        card_id: card_id.to_string(),
        x: fx + CARD_GAP + col as f64 * (CARD_W + CARD_GAP),
        y: fy + FRAME_LABEL_H + CARD_GAP + row as f64 * (CARD_H + CARD_GAP),
        w: CARD_W,
        h: CARD_H,
        z: 1,
    }
}
